src/librustc/infer/region_inference/mod.rs

```diff
@@ -253,6 +276,111 @@ pub struct RegionSnapshot {
     skolemization_count: u32,
 }
 
+#[derive(Debug)]
+pub struct TaintDirections {
+    incoming: bool,
+    outgoing: bool,
+}
+
+impl TaintDirections {
+    pub fn incoming() -> Self {
+        TaintDirections { incoming: true, outgoing: false }
+    }
+
+    pub fn outgoing() -> Self {
+        TaintDirections { incoming: false, outgoing: true }
+    }
+
+    pub fn both() -> Self {
+        TaintDirections { incoming: true, outgoing: true }
+    }
+}
+
+struct TaintSet {
```

Comment:
It'd be good to have a one-two sentence comment *somewhere* on this structure with a bit of explanation of its purpose.